src/Shared/WeakStringCache.cs

```diff
@@ -0,0 +1,355 @@
+        public string GetOrCreateEntry<T>(T internable, out bool cacheHit) where T : IInternable
+        {
+            int hashCode = GetInternableHashCode(internable);
+
+            StringBucket bucket;
+            string result;
+            lock (_stringsByHashCode)
+            {
+                if (!_stringsByHashCode.TryGetValue(hashCode, out bucket))
+                {
+                    bucket = new StringBucket();
+                }
+                result = bucket.GetString(internable);
+                if (result != null)
+                {
+                    cacheHit = true;
+                    return result;
+                }
+            }
+
+            // We don't have the string in the dictionary - create it.
+            result = internable.ExpensiveConvertToString();
+
+            lock (_stringsByHashCode)
+            {
+                bucket.AddString(result);
+
+                // Prevent the dictionary from growing forever with buckets whose underlying GC handles
+                // don't reference any live strings anymore.
+                if (_stringsByHashCode.Count >= _capacity)
+                {
+                    // Get rid of unused buckets.
+                    Scavenge();
+                    // And do this again if the number of buckets reaches double the current after-scavenge number.
+                    _capacity = _stringsByHashCode.Count * 2;
+                }
+                _stringsByHashCode[hashCode] = bucket;
+            }
+            cacheHit = false;
+            return result;
+        }
+
+        /// <summary>
+        /// Implements the simple yet very decently performing djb2 hash function.
+        /// </summary>
+        /// <param name="internable">The internable to compute the hash code for.</param>
+        /// <returns>The 32-bit hash code.</returns>
+        internal static int GetInternableHashCode<T>(T internable) where T : IInternable
+        {
+            int hashCode = 5381;
+            for (int i = 0; i < internable.Length; i++)
+            {
+                unchecked
+                {
+                    hashCode = hashCode * 33 + internable[i];
+                }
+            }
+            return hashCode;
+        }
+
+        /// <summary>
+        /// Iterates over the cache and removes unused buckets, i.e. buckets that don't reference live strings.
+        /// This is expensive so try to call such that the cost is amortized to O(1) per GetOrCreateEntry() invocation.
+        /// Assumes exclusive access to the dictionary, i.e. the lock is taken.
+        /// </summary>
+        public void Scavenge()
+        {
+            List<int> keysToRemove = null;
+            foreach (KeyValuePair<int, StringBucket> entry in _stringsByHashCode)
+            {
+                if (!entry.Value.IsUsed)
+                {
+                    entry.Value.Free();
+                    keysToRemove ??= new List<int>();
```

Comment:
Aside: in .NET Core you can safely remove keys from a `Dictionary<K,V>` while enumerating it. That' doesn't help you today of course.

Reply:
Nice, I didn't know! We could take advantage of that when building for Core but aside from the added complexity with `#ifdef`s I'm afraid that we currently target 2.1 and according to the documentation this is a 3.0+ feature. Maybe moving to 3.0 is OK at this point but it's going to be a separate change for sure.

Reply:
@rainersigwald do you still bits out of this branch destined to run on .NET Core 2.1?